// @salesforce/lwc-language-server/lib/metadata-utils/message-channel-util.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const glob_1 = require("glob");
const vscode_languageserver_1 = require("vscode-languageserver");
const util_1 = require("util");
const fs = __importStar(require("fs-extra"));
const glob = util_1.promisify(glob_1.Glob);
const MESSAGE_CHANNEL_DECLARATION_FILE = '.sfdx/typings/lwc/messagechannels.d.ts';
const MESSAGE_CHANNELS = new Set();
function resetMessageChannels() {
    MESSAGE_CHANNELS.clear();
}
exports.resetMessageChannels = resetMessageChannels;
function getResourceName(resourceMetaFile) {
    const resourceFile = resourceMetaFile.substring(0, resourceMetaFile.lastIndexOf('-'));
    return path_1.parse(resourceFile).name;
}
async function updateMessageChannelsIndex(updatedFiles, { workspaceRoots }, writeConfigs = true) {
    let didChange = false;
    for (const f of updatedFiles) {
        if (f.uri.endsWith('.messageChannel-meta.xml')) {
            if (f.type === vscode_languageserver_1.FileChangeType.Created) {
                didChange = true;
                MESSAGE_CHANNELS.add(getResourceName(f.uri));
            }
            else if (f.type === vscode_languageserver_1.FileChangeType.Deleted) {
                MESSAGE_CHANNELS.delete(getResourceName(f.uri));
                didChange = true;
            }
        }
    }
    if (didChange) {
        return processMessageChannels(workspaceRoots[0], writeConfigs);
    }
}
exports.updateMessageChannelsIndex = updateMessageChannelsIndex;
async function processMessageChannels(workspace, writeConfig) {
    if (MESSAGE_CHANNELS.size > 0 && writeConfig) {
        return fs.writeFile(path_1.join(workspace, MESSAGE_CHANNEL_DECLARATION_FILE), generateTypeDeclarations());
    }
}
async function indexMessageChannels(context, writeConfigs) {
    const { workspaceRoots } = context;
    const { sfdxPackageDirsPattern } = await context.getSfdxProjectConfig();
    const MESSAGE_CHANNEL_GLOB_PATTERN = `${sfdxPackageDirsPattern}/**/messageChannels/*.messageChannel-meta.xml`;
    try {
        const files = await glob(MESSAGE_CHANNEL_GLOB_PATTERN, { cwd: workspaceRoots[0] });
        for (const file of files) {
            MESSAGE_CHANNELS.add(getResourceName(file));
        }
        return processMessageChannels(workspaceRoots[0], writeConfigs);
    }
    catch (err) {
        console.log(`Error queuing up indexing of message channel resources. Error details:`, err);
        throw err;
    }
}
exports.indexMessageChannels = indexMessageChannels;
function generateTypeDeclarations() {
    let resTypeDecs = '';
    const sortedContentAssets = Array.from(MESSAGE_CHANNELS).sort();
    sortedContentAssets.forEach(res => {
        resTypeDecs += generateTypeDeclaration(res);
    });
    return resTypeDecs;
}
function generateTypeDeclaration(resourceName) {
    const result = `declare module "@salesforce/messageChannel/${resourceName}__c" {
    var ${resourceName}: string;
    export default ${resourceName};
}
`;
    return result;
}
