// @salesforce/source-deploy-retrieve/lib/client/deployStrategies/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const containerDeploy_1 = require("./containerDeploy");
const auraDeploy_1 = require("./auraDeploy");
const lwcDeploy_1 = require("./lwcDeploy");
const metadata_registry_1 = require("../../metadata-registry");
var constants_1 = require("./constants");
exports.AURA_TYPES = constants_1.AURA_TYPES;
exports.CONTAINER_ASYNC_REQUEST = constants_1.CONTAINER_ASYNC_REQUEST;
exports.METADATA_CONTAINER = constants_1.METADATA_CONTAINER;
var containerDeploy_2 = require("./containerDeploy");
exports.ContainerDeploy = containerDeploy_2.ContainerDeploy;
var auraDeploy_2 = require("./auraDeploy");
exports.AuraDeploy = auraDeploy_2.AuraDeploy;
var lwcDeploy_2 = require("./lwcDeploy");
exports.LwcDeploy = lwcDeploy_2.LwcDeploy;
exports.getDeployStrategy = (type, connection) => {
    switch (type) {
        case metadata_registry_1.registryData.types.auradefinitionbundle.name:
            return new auraDeploy_1.AuraDeploy(connection);
        case metadata_registry_1.registryData.types.lightningcomponentbundle.name:
            return new lwcDeploy_1.LwcDeploy(connection);
        default:
            return new containerDeploy_1.ContainerDeploy(connection);
    }
};
